core/DCache.hpp

```diff
@@ -26,29 +28,77 @@ namespace olympia
             PARAMETER(uint32_t, l1_associativity, 8, "DL1 associativity (power of 2)")
             PARAMETER(uint32_t, cache_latency, 1, "Assumed latency of the memory system")
             PARAMETER(bool, l1_always_hit, false, "DL1 will always hit")
+            PARAMETER(uint32_t, mshr_entries, 8, "Number of MSHR Entries")
         };
 
         static const char name[];
         DCache(sparta::TreeNode* n, const CacheParameterSet* p);
 
       private:
+        ////////////////////////////////////////////////////////////////////////////////
+        // L1 Cache Handling
+        ////////////////////////////////////////////////////////////////////////////////
+        using L1Handle = CacheFuncModel::Handle;
+        L1Handle l1_cache_;
+        const bool l1_always_hit_;
+        const uint32_t cache_latency_;
+        const uint64_t cache_line_size_;
+        const sparta::cache::AddrDecoderIF* addr_decoder_;
+        // Keep track of the instruction that causes current outstanding cache miss
+        //        MemoryAccessInfoPtr cache_pending_inst_ = nullptr;
+
+        const uint32_t num_mshr_entries_;
+
+        void setupL1Cache_(const CacheParameterSet* p);
+
         bool dataLookup_(const MemoryAccessInfoPtr & mem_access_info_ptr);
 
         void reloadCache_(uint64_t phy_addr);
 
-        void getInstsFromLSU_(const MemoryAccessInfoPtr & memory_access_info_ptr);
+        uint64_t getBlockAddr(const MemoryAccessInfoPtr & mem_access_info_ptr);
+
+        // To arbitrate between incoming request from LSU and Cache refills from BIU
+        //        bool incoming_cache_refill_ = false;
+        MemoryAccessInfoPtr incoming_cache_refill_ = nullptr;
+
+        using MSHREntryInfoPtr = sparta::SpartaSharedPointer<MSHREntryInfo>;
+        using MSHREntryIterator = sparta::Buffer<MSHREntryInfoPtr>::const_iterator;
+        // Ongoing Refill request
+        MSHREntryIterator current_refill_mshr_entry_;
+
+        // Cache Pipeline
+        enum class PipelineStage
+        {
+            LOOKUP = 0,
+            DATA_READ = 1,
+            DEALLOCATE = 2,
+            NUM_STAGES
+        };
+
+        sparta::Pipeline<MemoryAccessInfoPtr> cache_pipeline_{
+            "DCachePipeline", static_cast<uint32_t>(PipelineStage::NUM_STAGES), getClock()};
+
+        void handleLookup_();
+        void handleDataRead_();
+        void handleDeallocate_();
+
+        ////////////////////////////////////////////////////////////////////////////////
+        // Handle requests
+        ////////////////////////////////////////////////////////////////////////////////
+
+        void getMemReqFromLSU_(const MemoryAccessInfoPtr & memory_access_info_ptr);
 
         void getAckFromL2Cache_(const uint32_t & ack);
 
         void getRespFromL2Cache_(const MemoryAccessInfoPtr & memory_access_info_ptr);
```

Comment:
Comment within body for these get* methods says "Got..."; consider renaming to "receive". "get" also implies active requester whereas "receive" implies passive acceptor.